src/printer.js

```diff
@@ -2977,6 +2971,50 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+// A "leadingComment" attached to the top of the first token in the
+// return's argument, which would originally be in parens,
+// could trigger an ASI error where the parens are removed
+// and the comment bumped to the same line as the `return`,
+// while the contents of the return value are on the next line (oops!).
+//
+// This recurses the return argument, looking for the first token
+// (the leftmost leaf node) and, if it (or its parents) has any
+// leadingComments, wraps the return in parens.
+function returnArgumentHasLeadingComment(argument) {
+  const hasLeadingComment = (comments) => (
+    comments && comments.some(comment => comment.leading)
+  );
+  const getLeftSide = (node) => {
+    if (node.expressions) return node.expressions[0];
+    return node.left || node.test || node.callee || node.object || node.tag;
+  }
+  const NODE_TYPES_WITH_COMMENTABLE_LEFT_SIDE = {
+    "BinaryExpression": true,
+    "LogicalExpression": true,
+    "ConditionalExpression": true,
+    "CallExpression": true,
+    "MemberExpression": true,
+    "SequenceExpression": true,
+    "TaggedTemplateExpression": true,
+  };
+
+  if (hasLeadingComment(argument.comments)) return true;
+
+  if (NODE_TYPES_WITH_COMMENTABLE_LEFT_SIDE[argument.type]) {
```

Comment:
do you mind listing them as

```js
argument.type === "BinaryExpression" ||
argument.type === "LogicalExpression" ||
...
```

this way it's easier to manage. Feel free to put it in a variable: `hasCommentableLeftSide = ...`